docs/proposals/scheduling/binding-priority-preemption/README.md

```diff
@@ -0,0 +1,340 @@
+
+authors:
+- "@whitewindmills"
+
+reviewers:
+- "@RainbowMango"
+- "@XiShanYongYe-Chang"
+- "@chaunceyjiang"
+
+approvers:
+- "@RainbowMango"
+
+creation-date: 2024-05-27
+
+notes: This proposal refers to ResourceBinding and ClusterResourceBinding collectively as binding.
+---
+
+# Binding Priority and Preemption
+
+## Summary
+
+Currently, the scheduler only schedules workloads in order, that is, the workload triggered first can be scheduled first.
+And when a new binding has certain scheduling requirements that makes it infeasible on any member cluster,
+it may stay in the scheduling queue, until sufficient resources are free, and it can be scheduled.
+
+This proposal proposes the concept of priorities and preemption for bindings in Karmada and how the priority impacts scheduling and preemption of bindings
+when the member cluster runs out of resources.
+- The priority is independent of its workload type and must be one of the valid values, there is a total order on the values.
+- Preemption is the action taken when an important binding requires resources or conditions which are not available in all member clusters. So, one or more bindings need to be preempted to make room for the important binding.
+
+## Motivation
+
+It is fairly common in large-scale clusters to have more tasks than what Karmada can handle, so scheduling responses can become slow.
+Often times the workload is a mix of high priority critical tasks, and non-urgent tasks that can wait.
+Karmada should be able to distinguish these workloads in order to decide which ones should be scheduled sooner, which ones can wait and which ones can be preempted.
+
+### Goals
+
+- How to specify priority and preemption strategy for bindings.
+- Define how the order of these priorities are specified.
+- Define how new priority levels are added.
+- Effect of priority on scheduling and preemption.
+- Define the concept of binding preemption in Karmada.
+- Define scenarios under which a binding may get preempted.
+- Define mechanics of preemption.
+
+### Non-Goals
+
+- How preemption works in spread constraints.
+
+## Proposal
+
+### Terminology
+
+When a new binding has certain scheduling requirements that makes it infeasible on any member cluster,
+scheduler may choose to clear the scheduling result of lower priority bindings to satisfy the scheduling requirements of the new binding.
+We call this operation "binding preemption". Binding preemption is distinguished from "[policy preemption](../policy-preemption/README.md)" where high-priority policies preempt low-priority policies.
+
+### User Stories (Optional)
+
+#### As a user, I want to the task I submitted to be scheduled as soon as possible.
+
+I submit a workload that is prioritized above other workloads, but do not wish to discard existing work by preempting scheduled bindings.
+I hope that the high priority workload will be scheduled ahead of other queued bindings, as soon as sufficient cluster resources "naturally" become free.
```

Comment:
In case of a workload with higher priority can not be scheduled due to insufficient resources, will it block the others with lower priority?

1. If yes, it means no more workload could be scheduled until the one with the highest priority gets scheduled.
2. If not, the highest-priority workload might not be scheduled forever if low-priority workloads keep coming in.

Personally, I tend to option 1, which is exactly the user expectation I guess.

Reply:
I see @whitewindmills [answers below](https://github.com/karmada-io/karmada/blob/f1b5c033a3d06986de7b8c6ad88675fed62c2df4/docs/proposals/scheduling/binding-priority-preemption/README.md#effect-of-priority-on-scheduling):
> For example, a high priority binding is marked as unschedulable due to insufficient resources of member clusters, but the preemption behavior is not enabled. Karmada will try to schedule others with lower priority.

@LeonZh0u @wengyao04 What's your opinion on this?

Reply:
Hi, I think this makes sense. should we make it a feature gate (e.g. strict preemption mode)?
or we can have another grpc request type for estimate maxreplica that accounts for resources tied up in preemption at the moment.

Reply:
Yeah, we do need a feature gate. But here we are talking about the behavior.
I agree with @whitewindmills that without preemption if scheduling fails due to insufficient resources(or can not find feasible clusters), we should continue attempting to schedule other pending ResourcesBindings. This behavior aligns with he current implementation.